File: sdk/ts/src/receipt/idempotency.test.ts
import { describe, expect, it } from "vitest";
import { verifyChain } from "./chain.js";
import { createReceipt } from "./create.js";
import { canonicalize, hashReceipt } from "./hash.js";
import { agentReceiptSchema } from "./schema.js";
import { generateKeyPair, signReceipt } from "./signing.js";
import type { AgentReceipt } from "./types.js";

// buildChainWithKeys builds a signed, hash-linked chain whose i-th receipt
// carries idempotencyKeys[i] on action.idempotency_key (undefined = omitted).
function buildChainWithKeys(
	idempotencyKeys: (string | undefined)[],
	privateKey: string,
): AgentReceipt[] {
	const receipts: AgentReceipt[] = [];
	let previousHash: string | null = null;
	idempotencyKeys.forEach((key, i) => {
		const unsigned = createReceipt({
			issuer: { id: "did:agent:test" },
			principal: { id: "did:user:test" },
			action: {
				type: "filesystem.file.read",
				risk_level: "low",
				...(key !== undefined && { idempotency_key: key }),
			},
			outcome: { status: "success" },
			chain: {
				sequence: i + 1,
				previous_receipt_hash: previousHash,
				chain_id: "chain_test",
			},
		});
		const signed = signReceipt(unsigned, privateKey, "did:agent:test#key-1");
		receipts.push(signed);
		previousHash = hashReceipt(signed);
	});
	return receipts;
}

describe("idempotency_key", () => {
	it("createReceipt stamps action.idempotency_key", () => {
		const unsigned = createReceipt({
			issuer: { id: "did:agent:test" },
			principal: { id: "did:user:alice" },
			action: {
				type: "filesystem.file.read",
				risk_level: "low",
				idempotency_key: "req-1",
			},
			outcome: { status: "success" },
			chain: { sequence: 1, previous_receipt_hash: null, chain_id: "c" },
		});
		expect(unsigned.credentialSubject.action.idempotency_key).toBe("req-1");
	});

	it("omits idempotency_key from the canonical form when unset", () => {
		const unsigned = createReceipt({
			issuer: { id: "did:agent:test" },
			principal: { id: "did:user:alice" },
			action: { type: "filesystem.file.read", risk_level: "low" },
			outcome: { status: "success" },
			chain: { sequence: 1, previous_receipt_hash: null, chain_id: "c" },
		});
		expect(canonicalize(unsigned)).not.toContain("idempotency_key");
	});

	it("drops an empty idempotency_key instead of emitting it (spec §7.3.6)", () => {
		const unsigned = createReceipt({
			issuer: { id: "did:agent:test" },
			principal: { id: "did:user:alice" },
			action: {
				type: "filesystem.file.read",
				risk_level: "low",
				idempotency_key: "",
			},
			outcome: { status: "success" },
			chain: { sequence: 1, previous_receipt_hash: null, chain_id: "c" },
		});
		expect(unsigned.credentialSubject.action.idempotency_key).toBeUndefined();
		expect(canonicalize(unsigned)).not.toContain("idempotency_key");
	});

	it("rejects an empty idempotency_key in schema validation", () => {
		const { privateKey } = generateKeyPair();
		const unsigned = createReceipt({
			issuer: { id: "did:agent:test" },
			principal: { id: "did:user:alice" },
			action: { type: "filesystem.file.read", risk_level: "low" },
			outcome: { status: "success" },
			chain: { sequence: 1, previous_receipt_hash: null, chain_id: "c" },
		});
		const signed = signReceipt(unsigned, privateKey, "did:agent:test#key-1");
		const bad = {
			...signed,
			credentialSubject: {
				...signed.credentialSubject,
				action: { ...signed.credentialSubject.action, idempotency_key: "" },
			},
		};
		expect(agentReceiptSchema.safeParse(bad).success).toBe(false);
	});

	it("surfaces a duplicate idempotency_key as a warning, not a failure", () => {
		const { publicKey, privateKey } = generateKeyPair();
		// Receipts 0 and 2 share "req-A"; receipt 1 has a distinct key.
		const chain = buildChainWithKeys(["req-A", "req-B", "req-A"], privateKey);

		const result = verifyChain(chain, publicKey);
		expect(result.valid).toBe(true);
		expect(result.warnings).toHaveLength(1);
		expect(result.warnings?.[0]).toContain("req-A");
		expect(result.warnings?.[0]).toContain("0");
		expect(result.warnings?.[0]).toContain("2");
	});

	it("emits no warnings for distinct or absent keys", () => {
		const { publicKey, privateKey } = generateKeyPair();
		for (const keys of [
			["req-1", "req-2", "req-3"],
			[undefined, undefined, undefined],
			["req-1", undefined, "req-2"],
		]) {
			const chain = buildChainWithKeys(keys, privateKey);
			const result = verifyChain(chain, publicKey);
			expect(result.valid).toBe(true);
			expect(result.warnings).toBeUndefined();
		}
	});
});

File: sdk/go/receipt/idempotency_test.go
package receipt

import (
	"strings"
	"testing"
)

// buildChainWithKeys builds a signed, hash-linked chain whose i-th receipt
// carries idempotencyKeys[i] on action.idempotency_key (empty = omitted).
func buildChainWithKeys(t *testing.T, kp KeyPair, idempotencyKeys []string) []AgentReceipt {
	t.Helper()
	chain := make([]AgentReceipt, 0, len(idempotencyKeys))
	var prevHash *string
	for i, key := range idempotencyKeys {
		unsigned := Create(CreateInput{
			Issuer:    Issuer{ID: "did:agent:test"},
			Principal: Principal{ID: "did:user:test"},
			Action: Action{
				Type:           "filesystem.file.read",
				RiskLevel:      RiskLow,
				IdempotencyKey: key,
			},
			Outcome: Outcome{Status: StatusSuccess},
			Chain:   Chain{Sequence: i + 1, PreviousReceiptHash: prevHash, ChainID: "chain-1"},
		})
		signed, err := Sign(unsigned, kp.PrivateKey, "did:agent:test#key-1")
		if err != nil {
			t.Fatal(err)
		}
		chain = append(chain, signed)
		h, err := HashReceipt(signed)
		if err != nil {
			t.Fatal(err)
		}
		prevHash = &h
	}
	return chain
}

func TestCreateStampsIdempotencyKey(t *testing.T) {
	r := Create(CreateInput{
		Issuer:    Issuer{ID: "did:agent:test"},
		Principal: Principal{ID: "did:user:alice"},
		Action:    Action{Type: "filesystem.file.read", RiskLevel: RiskLow, IdempotencyKey: "req-1"},
		Outcome:   Outcome{Status: StatusSuccess},
		Chain:     Chain{Sequence: 1, ChainID: "chain-1"},
	})
	if r.CredentialSubject.Action.IdempotencyKey != "req-1" {
		t.Errorf("idempotency_key = %q, want %q", r.CredentialSubject.Action.IdempotencyKey, "req-1")
	}
}

// TestIdempotencyKeyOmittedWhenEmpty pins that an empty key serialises to an
// absent field (omitempty), so receipts that do not set it never carry it.
func TestIdempotencyKeyOmittedWhenEmpty(t *testing.T) {
	r := Create(CreateInput{
		Issuer:    Issuer{ID: "did:agent:test"},
		Principal: Principal{ID: "did:user:alice"},
		Action:    Action{Type: "filesystem.file.read", RiskLevel: RiskLow},
		Outcome:   Outcome{Status: StatusSuccess},
		Chain:     Chain{Sequence: 1, ChainID: "chain-1"},
	})
	canonical, err := Canonicalize(r)
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(canonical, "idempotency_key") {
		t.Errorf("canonical form should omit idempotency_key when empty: %s", canonical)
	}
}

// TestVerifyChainDuplicateIdempotencyKeyWarning verifies that a duplicate
// idempotency_key surfaces as a warning, not a failure (spec §7.3.6).
func TestVerifyChainDuplicateIdempotencyKeyWarning(t *testing.T) {
	kp, _ := GenerateKeyPair()
	// Receipts 0 and 2 share "req-A"; receipt 1 has a distinct key.
	chain := buildChainWithKeys(t, kp, []string{"req-A", "req-B", "req-A"})

	result := VerifyChain(chain, kp.PublicKey)
	if !result.Valid {
		t.Fatalf("duplicate idempotency_key must not invalidate the chain; broken at %d: %s", result.BrokenAt, result.Error)
	}
	if len(result.Warnings) != 1 {
		t.Fatalf("expected exactly 1 warning, got %d: %v", len(result.Warnings), result.Warnings)
	}
	w := result.Warnings[0]
	if !strings.Contains(w, "req-A") || !strings.Contains(w, "0") || !strings.Contains(w, "2") {
		t.Errorf("warning should name key req-A and indices 0, 2; got %q", w)
	}
}

// TestVerifyChainNoDuplicateNoWarning confirms that distinct (or absent) keys
// produce no warnings.
func TestVerifyChainNoDuplicateNoWarning(t *testing.T) {
	kp, _ := GenerateKeyPair()
	cases := map[string][]string{
		"all distinct": {"req-1", "req-2", "req-3"},
		"all absent":   {"", "", ""},
		"mixed absent": {"req-1", "", "req-2"},
	}
	for name, keys := range cases {
		t.Run(name, func(t *testing.T) {
			chain := buildChainWithKeys(t, kp, keys)
			result := VerifyChain(chain, kp.PublicKey)
			if !result.Valid {
				t.Fatalf("chain should be valid, broken at %d", result.BrokenAt)
			}
			if len(result.Warnings) != 0 {
				t.Errorf("expected no warnings, got %v", result.Warnings)
			}
		})
	}
}

// TestVerifyChainEmptyKeyNeverWarns pins that repeated *empty* keys (i.e.
// absent fields) are never treated as duplicates.
func TestVerifyChainEmptyKeyNeverWarns(t *testing.T) {
	kp, _ := GenerateKeyPair()
	chain := buildChainWithKeys(t, kp, []string{"", "", "", ""})
	result := VerifyChain(chain, kp.PublicKey)
	if len(result.Warnings) != 0 {
		t.Errorf("absent idempotency keys must never warn, got %v", result.Warnings)
	}
}
